core/consensus/pow/src/lib.rs

```diff
@@ -0,0 +1,408 @@
+) where
+	C: HeaderBackend<B> + AuxStore + 'static,
+	Algorithm: PowAlgorithm<B> + Send + Sync + 'static,
+	E: Environment<B> + Send + Sync + 'static,
+	E::Error: std::fmt::Debug,
+{
+	if let Err(_) = register_pow_inherent_data_provider(&inherent_data_providers) {
+		warn!("Registering inherent data provider for timestamp failed");
+	}
+
+	thread::spawn(move || {
+		loop {
+			match mine_loop(
+				&mut block_import,
+				client.as_ref(),
+				&algorithm,
+				&mut env,
+				&preruntime,
+				round,
+				&inherent_data_providers
+			) {
+				Ok(()) => (),
+				Err(e) => error!(
+					"Mining block failed with {:?}. Sleep for 1 second before restarting...",
+					e
+				),
+			}
+
+			std::thread::sleep(std::time::Duration::new(1, 0));
+		}
+	});
+}
+
+fn mine_loop<B: BlockT<Hash=H256>, C, Algorithm, E>(
+	block_import: &mut BoxBlockImport<B>,
+	client: &C,
+	algorithm: &Algorithm,
+	env: &mut E,
+	preruntime: &[u8],
+	round: u32,
+	inherent_data_providers: &inherents::InherentDataProviders,
+) -> Result<(), String> where
+	C: HeaderBackend<B> + AuxStore,
+	Algorithm: PowAlgorithm<B>,
+	E: Environment<B>,
+	E::Error: std::fmt::Debug,
+{
+	'outer: loop {
+		let best_hash = client.info().best_hash;
+		let best_header = client.header(BlockId::Hash(best_hash))
+			.map_err(|e| format!("Fetching best header failed: {:?}", e))?
+			.ok_or("Best header does not exist")?;
+		let mut aux = PowAux::read(client, &best_hash)?;
+		let mut proposer = env.init(&best_header).map_err(|e| format!("{:?}", e))?;
+
+		let inherent_data = inherent_data_providers
+			.create_inherent_data().map_err(String::from)?;
+		let mut inherent_digest = Digest::default();
+		inherent_digest.push(DigestItem::PreRuntime(POW_ENGINE_ID, preruntime.to_vec()));
+		let block = futures::executor::block_on(proposer.propose(
+			inherent_data,
+			inherent_digest,
+			std::time::Duration::new(0, 0)
+		)).map_err(|e| format!("Block proposing error: {:?}", e))?;
+
+		let (header, body) = block.deconstruct();
+		let seed = H256::random();
+		let (difficulty, seal) = {
+			loop {
+				let difficulty = algorithm.difficulty(
+					&BlockId::Hash(best_hash),
+				)?;
+
+				let seal = algorithm.mine(
+					&BlockId::Hash(best_hash),
+					&header.hash(),
+					&seed,
+					difficulty,
+					round,
+				)?;
+
+				if let Some(seal) = seal {
+					break (difficulty, seal)
+				}
+
+				if best_hash != client.info().best_hash {
+					continue 'outer
+				}
+			}
+		};
+
+		aux.total_difficulty = aux.total_difficulty.saturating_add(difficulty);
+		let hash = header.hash();
+
+		let key = aux_key(&hash);
+		let import_block = BlockImportParams {
+			origin: BlockOrigin::Own,
+			header,
+			justification: None,
+			post_digests: vec![DigestItem::Seal(POW_ENGINE_ID, seal)],
+			body: Some(body),
+			finalized: false,
+			auxiliary: vec![(key, Some(aux.encode()))],
+			fork_choice: ForkChoiceStrategy::Custom(true),
+		};
+
+		block_import.import_block(import_block, Default::default())
```

Comment:
Instead of `Default` it's better to write actual type that impls this trait. That way, the reader will be able to follow the logic without having to look at `import_block` prototype.